ghcide/session-loader/Development/IDE/Session.hs

```diff
@@ -611,19 +610,16 @@ loadSessionWithOptions recorder SessionLoadingOptions{..} dir = do
                                        [ "No cradle target found. Is this file listed in the targets of your cradle?"
                                        , "If you are using a .cabal file, please ensure that this module is listed in either the exposed-modules or other-modules section"
                                        ]
-
-          void $ modifyVar' fileToFlags $
-              Map.insert hieYaml this_flags_map
-          void $ modifyVar' filesMap $
-              flip HM.union (HM.fromList (map ((,hieYaml) . fst) $ concatMap toFlagsMap all_targets))
-
-          void $ extendKnownTargets all_targets
-
-          -- Invalidate all the existing GhcSession build nodes by restarting the Shake session
-          invalidateShakeCache
-
           -- The VFS doesn't change on cradle edits, re-use the old one.
-          restartShakeSession VFSUnmodified "new component" []
+          restartShakeSession VFSUnmodified "new component" [] $ do
+            void $ modifyVar' fileToFlags $
+                Map.insert hieYaml this_flags_map
+            void $ modifyVar' filesMap $
+                flip HM.union (HM.fromList (map ((,hieYaml) . fst) $ concatMap toFlagsMap all_targets))
+            key1 <- extendKnownTargets all_targets
```

Comment:
I am not sure running all these between sessions is a good idea @wz1000

Reply:
`fileToFlags` and `filesMap` don't need to updated in between sessions because they aren't part of the shake state, but I can't thinking that could go wrong by updating them here.

It is probably a good idea to call `extendKnownTargets` and `invalidateShakeCache` inside the callback though.

Reply:
I do have a performance concern, pushing too many things into the restart callback would slow down the build system by increasing its halting time.

Reply:
I think its fine to move the `modifyVar'` outside the callback.